pkg/storage/ingest/pusher.go

```diff
@@ -241,163 +252,281 @@ func (c pusherConsumer) shouldLogClientError(ctx context.Context, err error) (bo
 	}
+
+	p.wg.Wait()
+
 	return errs.Err()
 }
 
-type flushableWriteRequest struct {
-	*mimirpb.WriteRequest
-	context.Context
+// start starts the shards, each in its own goroutine.
+func (p *parallelStorageShards) start() {
+	shards := make([]*BatchingQueue, p.numShards)
+	p.wg.Add(p.numShards)
+
+	for i := range shards {
+		shards[i] = NewBatchingQueue(p.capacity, p.batchSize)
+		go p.run(shards[i])
+	}
+
+	p.shards = shards
 }
 
-func (p *shardingPusher) runShard(toFlush chan flushableWriteRequest) {
+// run runs the BatchingQueue for the shard.
+func (p *parallelStorageShards) run(queue *BatchingQueue) {
 	defer p.wg.Done()
-	for wr := range toFlush {
+	defer queue.Done()
+
+	for wr := range queue.Channel() {
 		p.numTimeSeriesPerFlush.Observe(float64(len(wr.WriteRequest.Timeseries)))
-		err := p.upstream.PushToStorage(wr.Context, wr.WriteRequest)
+		err := p.pusher.PushToStorage(wr.Context, wr.WriteRequest)
 		if err != nil {
-			p.errs <- err
+			queue.ErrorChannel() <- err
+		}
+	}
+}
+
+// BatchingQueue is a queue that batches the incoming time series according to the batch size.
+// Once the batch size is reached, the batch is pushed to a channel which can be accessed through the Channel() method.
+type BatchingQueue struct {
+	ch    chan FlushableWriteRequest
+	errCh chan error
+	done  chan struct{}
+
+	currentBatch FlushableWriteRequest
+	batchSize    int
+}
+
+// NewBatchingQueue creates a new BatchingQueue instance.
+func NewBatchingQueue(capacity int, batchSize int) *BatchingQueue {
+	return &BatchingQueue{
+		ch:           make(chan FlushableWriteRequest, capacity),
+		errCh:        make(chan error, capacity),
+		done:         make(chan struct{}),
+		currentBatch: FlushableWriteRequest{WriteRequest: &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()}},
+		batchSize:    batchSize,
+	}
+}
+
+// AddToBatch adds a time series to the current batch. If the batch size is reached, the batch is pushed to the Channel().
+func (q *BatchingQueue) AddToBatch(ctx context.Context, ts mimirpb.PreallocTimeseries) error {
+	s := &q.currentBatch
+	s.Timeseries = append(s.Timeseries, ts)
+	s.Context = ctx
+
+	if len(s.Timeseries) >= q.batchSize {
+		if err := q.push(*s); err != nil {
+			return err
+		}
+
+		q.currentBatch = FlushableWriteRequest{
+			WriteRequest: &mimirpb.WriteRequest{Timeseries: mimirpb.PreallocTimeseriesSliceFromPool()},
 		}
 	}
+
+	return nil
 }
 
-func (p *shardingPusher) close() error {
-	for shard, wr := range p.unfilledShards {
-		if len(wr.Timeseries) > 0 {
-			p.shards[shard] <- wr
+// Close closes the BatchingQueue, it'll push the current branch to the channel if it's not empty.
+// and then close the channel.
+func (q *BatchingQueue) Close() error {
+	if len(q.currentBatch.Timeseries) > 0 {
+		if err := q.push(q.currentBatch); err != nil {
+			return err
 		}
 	}
-	for _, shard := range p.shards {
-		close(shard)
+
+	close(q.ch)
+	<-q.done
+
+	select {
+	case err := <-q.errCh:
```

Comment:
depending on the capacity there may be more errors in this channel. So by closing early we might ignore some of them (some may be internal and may need to be retried for example).

So maybe we can either: 1) have errCh always with capacity 0 or 1; 2) loop over errCh to make sure we've drained the errors

I think I prefer 1) with capacity 0 because it matches the current implementation that we don't continue pushing if we've encountered an internal error. wdyt?

Reply:
Discussed this offline, we went a different direction by collecting all errors on either push or close and keeping the capacity as is.